### PR TITLE
Use jemalloc from vcpkg and built all the flavors

### PR DESCRIPTION
**Pull request by dcristoloveanu**
Use jemalloc from vcpkg and built all the flavors

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -27,6 +27,24 @@ option(run_reals_check "set run_reals_check to ON to run reals check (default is
 option(use_cppunittest "set use_cppunittest to ON to build CppUnitTest tests on Windows (default is OFF)" OFF)
 option(run_traceability "run traceability tool (default is ON)" ON)
 
+if((NOT "${CMAKE_BUILD_TYPE}" STREQUAL "RelWithDebInfo") AND (NOT "${CMAKE_BUILD_TYPE}" STREQUAL "Debug"))
+    message(FATAL_ERROR "CMAKE_BUILD_TYPE must be specified and be precisely one of Debug or RelWithDebInfo")
+endif()
```
Comment (anporumb) [inline CMakeLists.txt:32]: why is this needed?

Asking because I imagine this cmakelists.txt would "go down" in hierarchy all the way to c-pal/c-build-tools and it would fail there.

Comment (anporumb) [reply CMakeLists.txt:32]: this = all the yellow lines here.

File: vcpkg.json [hunk 2]
```diff
@@ -0,0 +1,9 @@
+{
+  "$schema": "https://raw.githubusercontent.com/microsoft/vcpkg-tool/main/docs/vcpkg.schema.json",
+  "name": "c-pal",
+  "version": "1.0.0",
```
Comment (anporumb) [inline vcpkg.json:4]: is there a c-pal package in vcpkg?

and it exists, who updates its version here?

**Review by anporumb — DISMISSED** [review]
:shipit:

File: CMakeLists.txt (hunk 1 above)
Comment (dcristoloveanu) [reply CMakeLists.txt]: I removed them. I was torn between explicitly enforcing this on each repo or relying on c-pal (in the future c-build-tools I guess) to do it. Because this should migrate to c-build-tools methinks.

File: vcpkg.json (hunk 2 above)
Comment (dcristoloveanu) [reply vcpkg.json:4]: We're using a dummy package name. It's not like we're going to make a push to include a port in vcpkga for it. Not soon at least.
Fixed to be com-wrapper at least

File: CMakeLists.txt [hunk 3]
```diff
@@ -43,6 +61,19 @@ set(run_unittests OFF)
...
 set(run_reals_check OFF)
 
+if ((WIN32) AND ("${GBALLOC_LL_TYPE}" STREQUAL "JEMALLOC"))
+    if (EXISTS ${CMAKE_CURRENT_LIST_DIR}/deps/vcpkg/scripts/buildsystems/vcpkg.cmake)
+        set(VCPKG_OVERLAY_TRIPLETS "${CMAKE_CURRENT_LIST_DIR}/triplets")
+        if(${fsanitize})
+            set(VCPKG_TARGET_TRIPLET "x64-windows-static-comwrapper-asan")
+        else()
+            set(VCPKG_TARGET_TRIPLET "x64-windows-static-comwrapper")
+        endif()
```
Comment (mattdurak) [inline CMakeLists.txt:53]: I think all this triplet stuff would be nice in c-build-tools. This whole block could be a function/macro from c-build-tools too so we don't have to do it in all the repos. But it sounds like moving things over there is the eventual plan anyway...

File: build/build_all.yml [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+jobs:
+- ${{ each GBALLOC_LL_TYPE in parameters.GBALLOC_LL_TYPE_VALUES }}:
+  - ${{ each FSANITIZE_TYPE in parameters.FSANITIZE_TYPE_VALUES }}:
+    - ${{ each ARCH_TYPE in parameters.ARCH_TYPE_VALUES }}:
+      - ${{ each BUILD_CONFIGURATION_TYPE in parameters.BUILD_CONFIGURATION_TYPE_VALUES }}:
+        # - ${{ if not( and( eq( GBALLOC_LL_TYPE , 'MIMALLOC' ), eq(FSANITIZE_TYPE, 'ON'))) }}: <==this fails to compile, reasons "2 parameters passed to if". The below line is equivalent.
+        # there is no point in running mimalloc and fsanitize, so that combination is excluded from the whole build.
+        - ${{ if or( ne( GBALLOC_LL_TYPE , 'JEMALLOC' ), ne(FSANITIZE_TYPE, 'ON')) }}:
+          - template: gated_jobs/build_and_run_tests.yml
```
Comment (mattdurak) [inline build/build_all.yml:35]: >build_and_run_tests [](http://example.com/codeflow?start=33&length=19)

I also wonder if this should be a template in c-build-tools for the general build/test pipelines where we don't need anything more advanced with service fabric/int tests/etc.

We already pull in yml templates from c-build-tools so it should work (in theory)

(just saying since it seems you're duplicating all of this right now anyway, but it's probably more of a task for later...)

**Review by mattdurak — APPROVED** [review]
:shipit:

**Review by anporumb — APPROVED** [review]
:shipit:

File: build/build_all.yml (hunk 4 above)
Comment (dcristoloveanu) [reply build/build_all.yml:35]: It should absolutely be in c-build-tools. And used in ctest, c-test-runnerswitcher and all.
